MIAUW-5614 - CodeStyle repository documentation (#37)
studyportals/CodeStyle: README.md

# CodeStyle
The purpose of this repository is to provide standardized configuration files for the most common linters we use in Studyportals repositories. The main usage is in [Super-Linter](https://github.com/studyportals/Super-Linter), which uses these configuration files to run the most important linters on most of our repositories.

Please send pull requests to Stefan Klokgieters or Rob van den Hout if you think you want to suggest changes.

## Table of Contents
- [Install](#install)
- [Linters](#linters)
  - [ESLint](#eslint)
  - [TSLint](#tslint)
  - [StyleLint](#stylelint)
  - [SCSSLint](#scsslint)
  - [PHP Code Sniffer](#php-code-sniffer)
  - [PHPMD](#phpmd)
- [Best Practices](#best-practices)
## Install
You can add CodeStyle as a dependency to your project by running the following command:

```bash
npm install @studyportals/code-style --save-dev
```

## Linters
### ESLint
[ESLint](https://eslint.org/) is a linter for JavaScript, as well as TypeScript code. It can help find and fix potential problems in your code.

> When you installed the CodeStyle package you can simply *extend* from the ESLint configuration file inside it.

Reviewer: ```suggestion
When you install the CodeStyle package you can simply *extend* from the ESLint configuration file inside it.
```

    }
}
```
#### Global variables
When global variables are used in different files than they we're defined in, ESLint will see it as an undefined variable. To solve this problem, you can define the global variables at the top of your file like so:
```javascript
/* global var1, var2, var3 */
```
 To avoid this you can define your project's global variables in a custom configuration.

 ### TSLint 
 *deprecated* - ESLint can now be used instead

> [TSLint](https://palantir.github.io/tslint/) is linter that is specifically meant to lint TypeScript code. When you installed the CodeStyle package you can simply *extend* from the TSLint configuration file inside it.

Reviewer: ```suggestion
[TSLint](https://palantir.github.io/tslint/) is a linter that is specifically meant to lint TypeScript code. When you install the CodeStyle package you can simply *extend* from the TSLint configuration file inside it.
```

```json
// tslint.json
{
    "extends": "./node_modules/@studyportals/code-style/tslint.json"
}
```

### StyleLint

> [StyleLint](https://stylelint.io/) is our main linter for all styling related files. It can read all types of style related syntax, like SCSS, Sass, Less and SugarSS. When you installed the CodeStyle package you can simply *extend* from the StyleLint configuration file inside it.

Reviewer: ```suggestion
[StyleLint](https://stylelint.io/) is our main linter for all styling related files. It can read all types of style related syntax, like SCSS, Sass, Less and SugarSS. When you install the CodeStyle package you can simply *extend* from the StyleLint configuration file inside it.
```

```json
// .stylelintrc.json
{
    "extends": "./node_modules/@studyportals/code-style/.stylelintrc.json"
}
```

### SCSSLint
*deprecated*

> [SCSSLint](https://github.com/sds/scss-lint) is a linter specifically meant for running against rulesets that are specific to `.scss` files. Generally though, we prefer to use [StyleLint](#stylelint) since it supports SCSS natively. When you installed the CodeStyle package you can simply reference the SCSSLint configuration file inside it from your custom configuration.

Reviewer: ```suggestion
[SCSSLint](https://github.com/sds/scss-lint) is a linter specifically meant for running against rulesets that are specific to `.scss` files. Generally though, we prefer to use [StyleLint](#stylelint) since it supports SCSS natively. When you install the CodeStyle package you can simply reference the SCSSLint configuration file inside it from your custom configuration.
```

```yaml
# .scss-lint.yml
options:
  config-file: ./node_modules/@studyportals/code-style/.scss-lint.yml
```

### PHP Code Sniffer

[PHP Code Sniffer](https://github.com/squizlabs/PHP_CodeSniffer) is a linter for PHP files. It can both detect and fix coding standard violations. PHP Code Sniffer can be run from the command line, where the configuration can be referenced.

```bash
phpcs-run --standard=./node_modules/@studyportals/code-style/phpcodesniffer.xml ./
```

### PHPMD

> [PHPMD](https://phpmd.org/) (or PHP Mess Detector) is a linter for PHP files. It takes a given PHP source code base and look for several potential problems within that source. When you installed the CodeStyle package you can reference the PHPMD configuration file inside it.

Reviewer: ```suggestion
[PHPMD](https://phpmd.org/) (or PHP Mess Detector) is a linter for PHP files. It takes a given PHP source code base and look for several potential problems within that source. When you install the CodeStyle package you can reference the PHPMD configuration file inside it.
```
